### PR TITLE
autotest support mlir

### PR DESCRIPTION
**Pull request by BBuf**
- 支持添加 `need-test-verbose` 的 PR label 让 CI 设置`ONEFLOW_TEST_VERBOSE`
- autotest的graph和mlir op测试可以完全共用同一套逻辑，这个pr做了一些修改包括设置check_mlir来enable mlir的round trip做dialect conversion以及outlinear来将代码主体作为外联函数执行。

当开启ONEFLOW_TEST_VERBOSE=1时，会在控制台输出oneflow dialect的IR文本方便查看。效果如下：

图片

本pr在多个op上进行了测试，暂未发现问题。

### CODE REVIEW

**Comment by hjchen2** [comment]
顺便支持一下xrt？

File: python/oneflow/test_utils/automated_test_util/torch_flow_dual_object.py [hunk 1]
```diff
@@ -852,13 +858,21 @@ def autotest(
...
 ):
     verbose = os.getenv("ONEFLOW_TEST_VERBOSE") is not None
 
     if check_graph == "ValidatedFlase":
         # check graph is intentionally closed and threre is a validated reason.
         check_graph = False
+    
+    if check_mlir:
+        os.environ["ONEFLOW_MLIR_ENABLE_ROUND_TRIP"] = "1"
+        os.environ["ONEFLOW_MLIR_ENABLE_CODEGEN_FUSERS"] = "1"
```
Comment (jackalcooper) [inline python/oneflow/test_utils/automated_test_util/torch_flow_dual_object.py:872]: 这个可以作为另外个标签吗

File: python/oneflow/test_utils/automated_test_util/torch_flow_dual_object.py [hunk 2]
```diff
@@ -852,13 +858,21 @@ def autotest(
     rtol=0.0001,
     atol=1e-05,
     check_graph=True,
+    check_mlir=True,
```
Comment (jackalcooper) [inline python/oneflow/test_utils/automated_test_util/torch_flow_dual_object.py:861]: 这里是不是可以考虑用是否开启MLIR编译来决定默认值？或者在之后判断一下。考虑到如果编译的时候没开启MLIR运行的时候会报错吗

Comment (BBuf) [reply python/oneflow/test_utils/automated_test_util/torch_flow_dual_object.py:861]: 不会报错，如果没有编译mlir，那么就只是走的nn.Graph测试。

File: python/oneflow/test_utils/automated_test_util/torch_flow_dual_object.py (hunk 1 above)
Comment (BBuf) [reply python/oneflow/test_utils/automated_test_util/torch_flow_dual_object.py:872]: ONEFLOW_MLIR_ENABLE_CODEGEN_FUSERS这个吗？

**Comment by BBuf** [comment]
> 顺便支持一下xrt？

可以，等我把这个pr改好了看看怎么一并支持。

File: python/oneflow/test_utils/automated_test_util/torch_flow_dual_object.py (hunk 1 above)
Comment (jackalcooper) [reply python/oneflow/test_utils/automated_test_util/torch_flow_dual_object.py:872]: 是的

Comment (strint) [reply python/oneflow/test_utils/automated_test_util/torch_flow_dual_object.py:872]: 主要是通过控制环境变量，在graph下验证了mlir相关执行逻辑？

Comment (jackalcooper) [reply python/oneflow/test_utils/automated_test_util/torch_flow_dual_object.py:872]: 嗯，主要考虑之后的构建job会被MLIR替换掉

Comment (BBuf) [reply python/oneflow/test_utils/automated_test_util/torch_flow_dual_object.py:872]: 我理解codegen_fuser在这里是没必要的，因为这里不能测试pattern，只能测试单个op，这个环境变量可以直接去掉？

Comment (jackalcooper) [reply python/oneflow/test_utils/automated_test_util/torch_flow_dual_object.py:872]: > 我理解codegen_fuser在这里是没必要的，因为这里不能测试pattern，只能测试单个op，这个环境变量可以直接去掉？

确实，去掉吧，想测试的话单独手动打开

File: oneflow/core/framework/nn_graph.cpp [hunk 3]
```diff
@@ -226,7 +226,7 @@ Maybe<void> NNGraph::RegisterNewVariableOpInJobPass() {
       CHECK_OR_RETURN(additional_variable_op_name_.insert(var_name).second)
           << " ERROR! variable Tensor with op_name: " << var_name
           << " has been add in nn.Graph: " << name_;
-    } else {
+    } else /* vairable ops from user code */ {
       CHECK_OR_RETURN(var_conf.initializer().has_empty_conf())
           << " nn.Graph ONLY support variable_op with empty conf,"
           << " because variable is inited by eager tensor."
```
Comment (strint) [inline oneflow/core/framework/nn_graph.cpp:232]: 这个两个check日志可以帮忙优化下说明。现在只说明了现象，没有说明主要的可能原因

Comment (BBuf) [reply oneflow/core/framework/nn_graph.cpp:232]: 盛航已refine。

**Comment by BBuf** [comment]
python3  /home/zhangxiaoyu/oneflow/python/oneflow/test/modules/test_add.py

```python
python3  /home/zhangxiaoyu/oneflow/python/oneflow/test/modules/test_add.py 
Environment has been initialized, this env init will do nothing.
/home/zhangxiaoyu/.local/lib/python3.8/site-packages/torch/_tensor.py:1013: UserWarning: The .grad attribute of a Tensor that is not a leaf Tensor is being accessed. Its .grad attribute won't be populated during autograd.backward(). If you indeed want the .grad field to be populated for a non-leaf Tensor, use .retain_grad() on the non-leaf Tensor. If you access the non-leaf Tensor by mistake, make sure you access the leaf Tensor instead. See github.com/pytorch/pytorch/pull/30531 for more informations. (Triggered internally at  aten/src/ATen/core/TensorBody.h:417.)
  return self._grad
.loc("broadcast_add-0":0:0): error: 'oneflow.broadcast_add' op operand #0 must be tensor of any type values, but got 'none'
loc("-":0:0): error: Failed to run round-trip passes
```

@jackalcooper

**Comment by BBuf** [comment]
> ```python
> broadcast_add
> ```

应该是scalar和一个Tensor相加触发了这个问题。

```python
x = 5
    y = flow.tensor(
        np.random.randn(*shape), requires_grad=True, device=flow.device(device)
    )
    of_out = flow.add(x, y).
```

**Comment by BBuf** [comment]
> 最小复现代码：
> 
> ```python
> import os
> 
> os.environ["ONEFLOW_MLIR_ENABLE_ROUND_TRIP"] = "1"
> 
> import oneflow as flow
> import oneflow.unittest
> 
> x = flow.tensor([10]).to(flow.bool)
> y = flow.tensor([10.0])
> x = x.to("cuda")
> y = y.to("cuda")
> 
> class GraphToRun(flow.nn.Graph):
>     def __init__(self):
>         super().__init__()
> 
>     def build(self, x, y):
>         res = x + y.mean().to(flow.bool)
>         # res.sum().backward()
>         return res
> 
> m = GraphToRun()
> m.debug(3)
> z = m(x, y)
> 
> print(z)
> ```

已解决。

**Comment by BBuf** [comment]
var op注册方式有问题，已改成op schema的方式，打开mlir 和 job的roundtrip已通过。

**Comment by BBuf** [comment]
constatpad op没有注册padding属性，已解决。

**Review by jackalcooper — APPROVED** [review]

**Review by Flowingsun007 — APPROVED** [review]

**Comment by BBuf** [comment]
```python
import os

os.environ["ONEFLOW_MLIR_ENABLE_ROUND_TRIP"] = "1"

import oneflow as flow
import oneflow.unittest

x = flow.tensor([10]).to(flow.bool)
x = x.to("cuda")

class GraphToRun(flow.nn.Graph):
    def __init__(self):
        super().__init__()

    def build(self, x):
        res = flow.ones_like(x)
        return res

m = GraphToRun()
z = m(x)
print(z)
```

最近一次的失败案例

**Comment by jackalcooper** [comment]
> 最近一次的失败案例

好的我来看看

**Comment by jackalcooper** [comment]
> 最近一次的失败案例

这个应该解决了

**Comment by BBuf** [comment]
在nn.Graph里面加了一个enable_mlir控制graph是否运行mlir，避免多机bug。本地测试可以生效。

@jackalcooper
